File: src/main/java/jatin/jatinapi/cc/bfhl.java
package jatin.jatinapi.cc;

import org.springframework.web.bind.annotation.*;
import java.util.*;

@RestController
public class bfhl {

    @PostMapping("/bfhl")
    public Map<String, Object> handleRequest(@RequestBody Map<String, Object> body) {

        Object result;

        if (body.containsKey("fibonacci")) {

            int n = ((Number) body.get("fibonacci")).intValue();
            result = generateFibSeries(n);

        } else if (body.containsKey("prime")) {

            List<Integer> input = (List<Integer>) body.get("prime");
            result = filterPrimes(input);

        } else if (body.containsKey("lcm")) {

            List<Integer> input = (List<Integer>) body.get("lcm");
            result = calculateLCM(input);
} else if (body.containsKey("hcf")) {

    List<Integer> input = (List<Integer>) body.get("hcf");
    result = calculateHCF(input);

}else if (body.containsKey("AI")) {

    String question = body.get("AI").toString();
    result = askAI(question);

}



        
        else {

            return Map.of(
                    "is_success", false,
                    "official_email", "jatin1256.be23@chitkarauniversity.edu.in",
                    "data", "Invalid key"
            );
        }

        return Map.of(
                "is_success", true,
                "official_email", "jatin1256.be23@chitkarauniversity.edu.in",
                "data", result
        );
    }


    private List<Integer> generateFibSeries(int n) {

        List<Integer> series = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            series.add(fib(i));
        }

        return series;
    }

    private int fib(int n) {

        if (n == 0) return 0;
        if (n == 1) return 1;

        return fib(n - 1) + fib(n - 2);
    }


    private List<Integer> filterPrimes(List<Integer> list) {

        List<Integer> primes = new ArrayList<>();

        for (int num : list) {
            if (isPrime(num)) {
                primes.add(num);
            }
        }

        return primes;
    }

    private boolean isPrime(int n) {

        if (n < 2) return false;

        for (int i = 2; i < n; i++) {
            if (n % i == 0) return false;
        }

        return true;
    }

    private int calculateLCM(List<Integer> list) {

    int lcm = list.get(0);

    for (int i = 1; i < list.size(); i++) {
        lcm = lcm(lcm, list.get(i));
    }

    return lcm;
}

private int lcm(int a, int b) {
    return (a * b) / gcd(a, b);
}


private int calculateHCF(List<Integer> list) {

    int hcf = list.get(0);

    for (int i = 1; i < list.size(); i++) {
        hcf = gcd(hcf, list.get(i));
    }

    return hcf;
}


private int gcd(int a, int b) {

    while (b != 0) {
        int temp = b;
        b = a % b;
        a = temp;
    }

    return a;
}

// ---------- AI ----------

private String askAI(String question) {

    try {

        String apiKey = "AIzaSyABl-G3RlvkcdGwjcBmnKJ7dY6cAydBqXk";

        String url = "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key=" + apiKey;

        org.springframework.web.reactive.function.client.WebClient client =
                org.springframework.web.reactive.function.client.WebClient.create();

        Map<String, Object> body = Map.of(
                "contents", List.of(
                        Map.of("parts", List.of(
                                Map.of("text", question+ "Answer in one word only")
                        ))
                )
        );

        Map response = client.post()
                .uri(url)
                .bodyValue(body)
                .retrieve()
                .bodyToMono(Map.class)
                .block();

        List candidates = (List) response.get("candidates");
        Map first = (Map) candidates.get(0);
        Map content = (Map) first.get("content");
        List parts = (List) content.get("parts");
        Map textPart = (Map) parts.get(0);

        return textPart.get("text").toString();

    } catch (Exception e) {

        return "AI error";

    }
}



}

File: api/bfhl.js
const fetch = require('node-fetch');

module.exports = async (req, res) => {
    // Enable CORS
    res.setHeader('Access-Control-Allow-Credentials', true);
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET,OPTIONS,PATCH,DELETE,POST,PUT');
    res.setHeader(
        'Access-Control-Allow-Headers',
        'X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version'
    );

    if (req.method === 'OPTIONS') {
        res.status(200).end();
        return;
    }

    if (req.method !== 'POST') {
        return res.status(405).json({ error: 'Method Not Allowed' });
    }

    try {
        const body = req.body;
        let result;

        if (body.fibonacci !== undefined) {
            const n = parseInt(body.fibonacci);
            result = generateFibSeries(n);
        } else if (body.prime !== undefined) {
            const input = body.prime;
            result = filterPrimes(input);
        } else if (body.lcm !== undefined) {
            const input = body.lcm;
            result = calculateLCM(input);
        } else if (body.hcf !== undefined) {
            const input = body.hcf;
            result = calculateHCF(input);
        } else if (body.AI !== undefined) {
            const question = body.AI.toString();
            result = await askAI(question);
        } else {
            return res.json({
                is_success: false,
                official_email: "jatin1256.be23@chitkarauniversity.edu.in",
                data: "Invalid key"
            });
        }

        return res.json({
            is_success: true,
            official_email: "jatin1256.be23@chitkarauniversity.edu.in",
            data: result
        });

    } catch (error) {
        console.error(error);
        return res.status(500).json({ is_success: false, error: "Internal Server Error" });
    }
};

function generateFibSeries(n) {
    const series = [];
    for (let i = 0; i < n; i++) {
        series.push(fib(i));
    }
    return series;
}

function fib(n) {
    if (n === 0) return 0;
    if (n === 1) return 1;
    let a = 0, b = 1;
    for (let i = 2; i <= n; i++) {
        let temp = a + b;
        a = b;
        b = temp;
    }
    return b; // Note: original recursive, iterative is safer for JS stack
    // But to match exact logic:
    // if (n === 0) return 0;
    // if (n === 1) return 1;
    // return fib(n - 1) + fib(n - 2); 
    // Optimization: The original used recursion which is slow. I'll use the iterative one above which is correct.
}

// Re-implementing recursive to be 100% faithful to "logic", but optimization is better.
// Actually, let's use the recursive one if n is small, but for safety in Node (single threaded), iterative is better.
// Let's stick to the efficient implementation.

function filterPrimes(list) {
    const primes = [];
    for (const num of list) {
        if (isPrime(num)) {
            primes.push(num);
        }
    }
    return primes;
}

function isPrime(n) {
    if (n < 2) return false;
    for (let i = 2; i < n; i++) {
        if (n % i === 0) return false;
    }
    return true;
}

function calculateLCM(list) {
    if (!list || list.length === 0) return 0;
    let result = list[0];
    for (let i = 1; i < list.length; i++) {
        result = lcm(result, list[i]);
    }
    return result;
}

function lcm(a, b) {
    if (a === 0 || b === 0) return 0;
    return Math.abs(a * b) / gcd(a, b);
}

function calculateHCF(list) {
    if (!list || list.length === 0) return 0;
    let result = list[0];
    for (let i = 1; i < list.length; i++) {
        result = gcd(result, list[i]);
    }
    return result;
}

function gcd(a, b) {
    while (b !== 0) {
        let temp = b;
        b = a % b;
        a = temp;
    }
    return a;
}

async function askAI(question) {
    try {
        const apiKey = "AIzaSyABl-G3RlvkcdGwjcBmnKJ7dY6cAydBqXk";
        const url = `https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key=${apiKey}`;

        const body = {
            contents: [{
                parts: [{
                    text: question + "Answer in one word only"
                }]
            }]
        };

        const response = await fetch(url, {
            method: 'POST',
            body: JSON.stringify(body),
            headers: { 'Content-Type': 'application/json' }
        });

        const data = await response.json();
        
        if (data.candidates && data.candidates.length > 0) {
             return data.candidates[0].content.parts[0].text;
        }
        return "No response from AI";

    } catch (e) {
        return "AI error";
    }
}
